packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -0,0 +1,87 @@
+import type { LoadedStrapi } from '@strapi/types';
+import { omit } from 'lodash/fp';
+import { HISTORY_VERSION_UID } from '../constants';
+
+import type { HistoryVersions } from '../../../../shared/contracts';
+
+const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
+  /**
+   * Use the query engine API, not the document service,
+   * since we'll refactor history version to be just a model instead of a content type.
+   * TODO: remove this comment once the refactor is done.
+   */
+  const query = strapi.db.query(HISTORY_VERSION_UID);
+
+  let isInitialized = false;
+
+  return {
+    async init() {
+      // Prevent initializing the service twice
+      if (isInitialized) {
+        return;
+      }
+
+      // TODO: replace by strapi.documents.use once it supports multiple actions at once
+      strapi.documents?.middlewares.add('_all', '_all', async (context, next) => {
+        // Ignore actions that don't mutate documents
+        if (!['create', 'update', 'publish', 'unpublish'].includes(context.action)) {
+          return next(context);
+        }
+
+        // Ignore content types not created by the user
+        if (!context.uid.startsWith('api::')) {
+          return next(context);
+        }
+
+        const fieldsToIgnore = [
+          'createdAt',
+          'updatedAt',
+          'publishedAt',
+          'createdBy',
+          'updatedBy',
+          'localizations',
+          'locale',
+          'strapi_stage',
+          'strapi_assignee',
+        ];
+
+        /**
+         * Await the middleware stack because for create actions,
+         * the document ID only exists after the creation, which is later in the stack.
+         */
+        const result = await next(context);
+
+        // Prevent creating a history version for an action that wasn't actually executed
+        await strapi.db.transaction(async ({ onCommit }) => {
+          onCommit(() => {
+            this.createVersion({
+              contentType: context.uid,
+              relatedDocumentId: result.documentId,
+              locale: context.params.locale,
+              // TODO: check if drafts should should be "modified" once D&P is ready
+              status: context.params.status,
+              data: omit(fieldsToIgnore, context.params.data),
+              schema: omit(fieldsToIgnore, strapi.contentType(context.uid).attributes),
+            });
+          });
+        });
+
+        return result;
+      });
+
+      isInitialized = true;
+    },
+
+    async createVersion(historyVersionData: HistoryVersions.CreateHistoryVersion) {
+      await query.create({
+        data: {
+          ...historyVersionData,
+          createdAt: new Date(),
+          createdBy: strapi.requestContext.get()?.state?.user.id,
```

Comment:
Already discussed on slack but put this here anyway. I don't know how it works now but in v4 these were already set when using content-types, well you get them using an util:

https://github.com/strapi/strapi/blob/f114e594550e18411b8fa6e456570aca868ae356/packages/core/content-releases/server/src/services/release.ts#L53